src/node/internal/internal_timers_promises.ts

```diff
@@ -62,16 +64,23 @@ export async function setTimeout<T = void>(
   }
 
   const { promise, resolve, reject } = Promise.withResolvers<T>();
+  let onCancel: OnCancelCallback;
 
   const timer = timers.setTimeout(() => {
     resolve(value as T);
+    if (onCancel) {
+      signal?.removeEventListener('abort', onCancel);
+    }
   }, delay ?? 0);
 
   if (signal) {
-    function onCancel(): void {
+    onCancel = (): void => {
       timers.clearTimeout(timer);
-      reject(new AbortError(undefined, { cause: signal?.reason }));
-    }
+      if (onCancel) {
+        signal.removeEventListener('abort', onCancel);
+      }
```

Comment:
this code can only be excuted when onCancel is defined
```suggestion
      signal.removeEventListener('abort', onCancel);
```

Reply:
This won't work because it can be undefined if no signal is provided. Changing this line requires adding non-null assertion, which also requires adding a eslint-ignore-next-line comment